File: grendel/calypso/util/StringUtils.java
/* -*- Mode: Java; tab-width: 2; indent-tabs-mode: nil -*-
 *
 * The contents of this file are subject to the Mozilla Public
 * License Version 1.1 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of
 * the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS
 * IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
 * implied. See the License for the specific language governing
 * rights and limitations under the License.
 *
 * The Original Code is the Grendel mail/news client.
 *
 * The Initial Developer of the Original Code is Netscape Communications
 * Corporation.  Portions created by Netscape are
 * Copyright (C) 1997 Netscape Communications Corporation. All
 * Rights Reserved.
 *
 * Contributor(s): 
 */

package calypso.util;

public class StringUtils {
  /**
   * Uppercase the characters in aString.
   *
   * @see: java.lang.Character
   */
  static public String UpperCase(String aString) {
    StringBuf buf = StringBuf.Alloc();
    buf.append(aString);
    buf.toUpperCase();
    aString = buf.toString();
    StringBuf.Recycle(buf);
    return aString;
  }

  /**
   * Lowercase the characters in aString.
   *
   * @see: java.lang.Character
   */
  static public String LowerCase(String aString) {
    StringBuf buf = StringBuf.Alloc();
    buf.append(aString);
    buf.toLowerCase();
    aString = buf.toString();
    StringBuf.Recycle(buf);
    return aString;
  }

  /**
   * Return true if the string buffer contains nothing but whitespace
   * as defined by Character.isWhitespace()
   *
   * @see: java.lang.Character
   */
  static public boolean IsWhitespace(String aString) {
    // XXX this should just loop right here instead of copying
    // the string because the loop will be inlined with a good
    // compiler
    StringBuf buf = StringBuf.Alloc();
    buf.append(aString);
    boolean rv = buf.isWhitespace();
    StringBuf.Recycle(buf);
    return rv;
  }

  /**
   * Translate an integer into a string that is at least aDigits
   * wide. Pad with zeros if necessary.
   */
  static public String ToHex(int i, int aDigits) {
    String rv = Integer.toString(i, 16);
    int len = rv.length();
    if (len < aDigits) {
      StringBuf buf = StringBuf.Alloc();
      len = aDigits - len;
      while (--len >= 0) {
        buf.append('0');
      }
      buf.append(rv);
      rv = buf.toString();
      StringBuf.Recycle(buf);
    }
    return rv;
  }

  /**
   * Compress the whitespace out of a string and return a new string
   */
  static public String CompressWhitespace(String aString, boolean aLeading) {
    StringBuf buf = StringBuf.Alloc();
    buf.append(aString);
    buf.compressWhitespace(aLeading);
    String rv = buf.toString();
    StringBuf.Recycle(buf);
    return rv;
  }

  /**
   * Quote a string using java source file rules. The result is a new
   * string with all the appropriate data quoted. The outer quotes that
   * would be required in a java source file are not provided by this
   * routine unless aProvideOuterQuotes is true.
   */
  static char[] hex;
  static {
    hex = new char[16];
    "0123456789abcdef".getChars(0, 16, hex, 0);
  }
  static public String JavaQuote(String aString, boolean aProvideOuterQuotes)
  {
    StringBuf buf = StringBuf.Alloc();
    if (aProvideOuterQuotes) {
      buf.append('"');
    }
    for (int i = 0, n = aString.length(); i < n; i++) {
      char ch = aString.charAt(i);
      switch (ch) {
      case '\n':
        buf.append("\\n");
        break;
      case '\r':
        buf.append("\\r");
        break;
      case '\t':
        buf.append("\\t");
        break;
      case '"':
        buf.append("\\\"");
        break;
      default:
        if ((ch < 32) || (ch >= 127)) {
          buf.append("\\u");
          buf.append(hex[(ch >> 12) & 0xf]);
          buf.append(hex[(ch >> 8) & 0xf]);
          buf.append(hex[(ch >> 4) & 0xf]);
          buf.append(hex[ch & 0xf]);
        } else {
          buf.append(ch);
        }
      }
    }
    if (aProvideOuterQuotes) {
      buf.append('"');
    }
    String rv = buf.toString();
    StringBuf.Recycle(buf);
    return rv;
  }
}

File: plugin/oji/MRJ/plugin/Source/StringUtils.cpp
/*
 * The contents of this file are subject to the Netscape Public License
 * Version 1.0 (the "NPL"); you may not use this file except in
 * compliance with the NPL.  You may obtain a copy of the NPL at
 * http://www.mozilla.org/NPL/
 *
 * Software distributed under the NPL is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the NPL
 * for the specific language governing rights and limitations under the
 * NPL.
 *
 * The Initial Developer of this code under the NPL is Netscape
 * Communications Corporation.  Portions created by Netscape are
 * Copyright (C) 1998 Netscape Communications Corporation.  All Rights
 * Reserved.
 */

/*
	StringUtils.cpp
 */

#include "StringUtils.h"

#include <string.h>

inline unsigned char toupper(unsigned char c)
{
	return (c >= 'a' && c <= 'z') ? (c - ('a' - 'A')) : c;
}

int strcasecmp(const char * str1, const char * str2)
{
#if !__POWERPC__
	
	const	unsigned char * p1 = (unsigned char *) str1;
	const	unsigned char * p2 = (unsigned char *) str2;
				unsigned char		c1, c2;
	
	while (toupper(c1 = *p1++) == toupper(c2 = *p2++))
		if (!c1)
			return(0);

#else
	
	const	unsigned char * p1 = (unsigned char *) str1 - 1;
	const	unsigned char * p2 = (unsigned char *) str2 - 1;
				unsigned long		c1, c2;
		
	while (toupper(c1 = *++p1) == toupper(c2 = *++p2))
		if (!c1)
			return(0);

#endif
	
	return(toupper(c1) - toupper(c2));
}

char* strdup(const char* str)
{
	if (str != NULL) {
		char* result = new char[::strlen(str) + 1];
		if (result != NULL)
			::strcpy(result, str);
		return result;
	}
	return NULL;
}
